### PR TITLE
r/fsx_windows_file_system - add `audit_log_configuration`

### LINKED ISSUE #19793: aws_fsx_windows_file_system sends all self_managed_active_directory settings if any setting in that block has changed
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

$ terraform -v
Terraform v0.13.5
+ provider registry.terraform.io/hashicorp/archive v2.2.0
+ provider registry.terraform.io/hashicorp/aws v3.45.0
+ provider registry.terraform.io/hashicorp/template v2.2.0

### Affected Resource(s)

* aws_fsx_windows_file_system

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
/*
Imported FSX Filesystem
*/
resource "aws_fsx_windows_file_system" "fsx1" {
  deployment_type     = "MULTI_AZ_1"
  storage_capacity    = 13312
  subnet_ids          = [module.vpc.private_subnets[0], module.vpc.private_subnets[1]]
  throughput_capacity = 128
  security_group_ids  = var.fsx_sg_ids

  /*
  Changes to self_managed_active_directory block are better done through console and just update them in the tf file for imported filesystems
  */
  self_managed_active_directory {
    dns_ips                                = var.fsx_dns_ips
    domain_name                            = var.fsx_domain_name
    password                               = var.fsx_ad_sa_password
    username                               = var.fsx_ad_sa_name
    organizational_unit_distinguished_name = var.fsx_ou_dn
    file_system_administrators_group       = "AWS Corp WorkSpaces Prod FSX1 Admins"
  }
  tags = {
    Name = "corp-workspaces-prod-fsx1"
  }
  /*
  Certain resource arguments, like security_group_ids and the self_managed_active_directory configuation block password, do not have a FSx API method for reading the information after creation. If these arguments are set in the Terraform configuration on an imported resource, Terraform will always show a difference. To workaround this behavior, either omit the argument from the Terraform configuration or use ignore_changes to hide the difference
  */
  lifecycle {
    prevent_destroy = true
    ignore_changes = [
      security_group_ids,
      self_managed_active_directory.0.password,
    ]
  }
}
```

### Expected Behavior

When updating self_managed_active_directory.0.dns_ips (for example), only that field should be sent to AWS in the update request.

### Actual Behavior

When updating self_managed_active_directory.0.dns_ips the provider also sent the username and password along with it even though the password had not changed in the code.  As this is an imported filesystem the password in state is empty and this led to a validation error. 

### Steps to Reproduce

1. Import a manually created filesystem
2. Update dns_ips for the filesystem in Terraform
3. `terraform apply`

Alternatively if the filesystem is created in TF and a manual rotation of the password is performed, you can validate that the password gets reverted by TF.

### Important Factoids

This appears to be a bug in `expandFsxSelfManagedActiveDirectoryConfigurationUpdate` which is copying the fields into the update struct without verifying that each individual field has actually changed:

```go
req := &fsx.SelfManagedActiveDirectoryConfigurationUpdates{
		DnsIps:   expandStringSet(data["dns_ips"].(*schema.Set)),
		Password: aws.String(data["password"].(string)),
		UserName: aws.String(data["username"].(string)),
	}
```

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #19793

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSFsxWindowsFileSystem_'
--- PASS: TestAccAWSFsxWindowsFileSystem_auditConfig (1458.07s)
```

### CODE REVIEW

File: aws/resource_aws_fsx_windows_file_system.go [hunk 1]
```diff
@@ -45,6 +47,38 @@ func resourceAwsFsxWindowsFileSystem() *schema.Resource {
...
 				Computed: true,
 			},
+			"audit_log_configuration": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"audit_log_destionation": {
```
Comment (ewbankkit) [inline aws/resource_aws_fsx_windows_file_system.go:57]: Typo.

```suggestion
						"audit_log_destination": {
```

File: aws/resource_aws_fsx_windows_file_system.go [hunk 2]
```diff
@@ -522,3 +562,53 @@ func flattenFsxSelfManagedActiveDirectoryConfiguration(d *schema.ResourceData, a
...
+		return nil
+	}
+
+	data := l[0].(map[string]interface{})
+	req := &fsx.WindowsAuditLogCreateConfiguration{
+		FileAccessAuditLogLevel:      aws.String(data["file_access_audit_log_level"].(string)),
+		FileShareAccessAuditLogLevel: aws.String(data["file_share_access_audit_log_level"].(string)),
+	}
+
+	if v, ok := data["audit_log_destionation"].(string); ok && v != "" {
```
Comment (ewbankkit) [inline aws/resource_aws_fsx_windows_file_system.go:577]:
```suggestion
	if v, ok := data["audit_log_destination"].(string); ok && v != "" {
```

File: aws/resource_aws_fsx_windows_file_system.go [hunk 3]
```diff
@@ -522,3 +562,53 @@ func flattenFsxSelfManagedActiveDirectoryConfiguration(d *schema.ResourceData, a
...
+		return []map[string]interface{}{}
+	}
+
+	m := map[string]interface{}{
+		"file_access_audit_log_level":       aws.StringValue(adopts.FileAccessAuditLogLevel),
+		"file_share_access_audit_log_level": aws.StringValue(adopts.FileShareAccessAuditLogLevel),
+	}
+
+	if adopts.AuditLogDestination != nil {
+		m["audit_log_destionation"] = aws.StringValue(adopts.AuditLogDestination)
```
Comment (ewbankkit) [inline aws/resource_aws_fsx_windows_file_system.go:595]:
```suggestion
		m["audit_log_destination"] = aws.StringValue(adopts.AuditLogDestination)
```

File: aws/resource_aws_fsx_windows_file_system_test.go [hunk 4]
```diff
@@ -702,6 +706,50 @@ func TestAccAWSFsxWindowsFileSystem_WeeklyMaintenanceStartTime(t *testing.T) {
...
+		CheckDestroy: testAccCheckFsxWindowsFileSystemDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAwsFsxWindowsFileSystemConfigAuditConfig(rName, "SUCCESS_ONLY"),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckFsxWindowsFileSystemExists(resourceName, &filesystem),
+					resource.TestCheckResourceAttr(resourceName, "audit_log_configuration.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "audit_log_configuration.0.file_access_audit_log_level", "SUCCESS_ONLY"),
+					resource.TestCheckResourceAttr(resourceName, "audit_log_configuration.0.file_share_access_audit_log_level", "SUCCESS_ONLY"),
+					resource.TestCheckResourceAttrSet(resourceName, "audit_log_configuration.0.audit_log_destionation"),
```
Comment (ewbankkit) [inline aws/resource_aws_fsx_windows_file_system_test.go:727]:
```suggestion
					resource.TestCheckResourceAttrSet(resourceName, "audit_log_configuration.0.audit_log_destination"),
```

File: aws/resource_aws_fsx_windows_file_system_test.go [hunk 5]
```diff
@@ -702,6 +706,50 @@ func TestAccAWSFsxWindowsFileSystem_WeeklyMaintenanceStartTime(t *testing.T) {
...
+				},
+			},
+			{
+				Config: testAccAwsFsxWindowsFileSystemConfigAuditConfig(rName, "SUCCESS_AND_FAILURE"),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckFsxWindowsFileSystemExists(resourceName, &filesystem),
+					resource.TestCheckResourceAttr(resourceName, "audit_log_configuration.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "audit_log_configuration.0.file_access_audit_log_level", "SUCCESS_AND_FAILURE"),
+					resource.TestCheckResourceAttr(resourceName, "audit_log_configuration.0.file_share_access_audit_log_level", "SUCCESS_AND_FAILURE"),
+					resource.TestCheckResourceAttrSet(resourceName, "audit_log_configuration.0.audit_log_destionation"),
```
Comment (ewbankkit) [inline aws/resource_aws_fsx_windows_file_system_test.go:746]:
```suggestion
					resource.TestCheckResourceAttrSet(resourceName, "audit_log_configuration.0.audit_log_destionation"),
```
```suggestion
					resource.TestCheckResourceAttrSet(resourceName, "audit_log_configuration.0.audit_log_destination"),
```

File: aws/resource_aws_fsx_windows_file_system_test.go [hunk 6]
```diff
@@ -1129,3 +1177,25 @@ resource "aws_fsx_windows_file_system" "test" {
...
+
+resource "aws_fsx_windows_file_system" "test" {
+  active_directory_id = aws_directory_service_directory.test.id
+  skip_final_backup   = true
+  storage_capacity    = 32
+  subnet_ids          = [aws_subnet.test1.id]
+  throughput_capacity = 32
+
+  audit_log_configuration {
+    audit_log_destionation            = aws_cloudwatch_log_group.test.arn
```
Comment (ewbankkit) [inline aws/resource_aws_fsx_windows_file_system_test.go:1195]:
```suggestion
    audit_log_destination            = aws_cloudwatch_log_group.test.arn
```

File: website/docs/r/fsx_windows_file_system.html.markdown [hunk 7]
```diff
@@ -80,6 +81,12 @@ The following arguments are supported for `self_managed_active_directory` config
...
 
+### audit_log_configuration
+
+* `audit_log_destionation` - (Optional) The Amazon Resource Name (ARN) for the destination of the audit logs. The destination can be any Amazon CloudWatch Logs log group ARN or Amazon Kinesis Data Firehose delivery stream ARN. Can be specified when `file_access_audit_log_level` and `file_share_access_audit_log_level` are not set to `DISABLED`. The name of the Amazon CloudWatch Logs log group must begin with the `/aws/fsx` prefix. The name of the Amazon Kinesis Data Firehouse delivery stream must begin with the `aws-fsx` prefix. If you do not provide a destination in `audit_log_destionation`, Amazon FSx will create and use a log stream in the CloudWatch Logs /aws/fsx/windows log group.
```
Comment (ewbankkit) [inline website/docs/r/fsx_windows_file_system.html.markdown:86]:
```suggestion
* `audit_log_destination` - (Optional) The Amazon Resource Name (ARN) for the destination of the audit logs. The destination can be any Amazon CloudWatch Logs log group ARN or Amazon Kinesis Data Firehose delivery stream ARN. Can be specified when `file_access_audit_log_level` and `file_share_access_audit_log_level` are not set to `DISABLED`. The name of the Amazon CloudWatch Logs log group must begin with the `/aws/fsx` prefix. The name of the Amazon Kinesis Data Firehouse delivery stream must begin with the `aws-fsx` prefix. If you do not provide a destination in `audit_log_destionation`, Amazon FSx will create and use a log stream in the CloudWatch Logs /aws/fsx/windows log group.
```

**Comment by DrFaust92** [comment]
Amended the typo 🤦

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSFsxWindowsFileSystem_'                                 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSFsxWindowsFileSystem_ -timeout 180m
=== RUN   TestAccAWSFsxWindowsFileSystem_basic
=== PAUSE TestAccAWSFsxWindowsFileSystem_basic
=== RUN   TestAccAWSFsxWindowsFileSystem_singleAz2
=== PAUSE TestAccAWSFsxWindowsFileSystem_singleAz2
=== RUN   TestAccAWSFsxWindowsFileSystem_storageTypeHdd
=== PAUSE TestAccAWSFsxWindowsFileSystem_storageTypeHdd
=== RUN   TestAccAWSFsxWindowsFileSystem_multiAz
=== PAUSE TestAccAWSFsxWindowsFileSystem_multiAz
=== RUN   TestAccAWSFsxWindowsFileSystem_disappears
=== PAUSE TestAccAWSFsxWindowsFileSystem_disappears
=== RUN   TestAccAWSFsxWindowsFileSystem_AutomaticBackupRetentionDays
=== PAUSE TestAccAWSFsxWindowsFileSystem_AutomaticBackupRetentionDays
=== RUN   TestAccAWSFsxWindowsFileSystem_CopyTagsToBackups
=== PAUSE TestAccAWSFsxWindowsFileSystem_CopyTagsToBackups
=== RUN   TestAccAWSFsxWindowsFileSystem_DailyAutomaticBackupStartTime
=== PAUSE TestAccAWSFsxWindowsFileSystem_DailyAutomaticBackupStartTime
=== RUN   TestAccAWSFsxWindowsFileSystem_KmsKeyId
=== PAUSE TestAccAWSFsxWindowsFileSystem_KmsKeyId
=== RUN   TestAccAWSFsxWindowsFileSystem_SecurityGroupIds
=== PAUSE TestAccAWSFsxWindowsFileSystem_SecurityGroupIds
=== RUN   TestAccAWSFsxWindowsFileSystem_SelfManagedActiveDirectory
=== PAUSE TestAccAWSFsxWindowsFileSystem_SelfManagedActiveDirectory
=== RUN   TestAccAWSFsxWindowsFileSystem_SelfManagedActiveDirectory_Username
=== PAUSE TestAccAWSFsxWindowsFileSystem_SelfManagedActiveDirectory_Username
=== RUN   TestAccAWSFsxWindowsFileSystem_StorageCapacity
=== PAUSE TestAccAWSFsxWindowsFileSystem_StorageCapacity
=== RUN   TestAccAWSFsxWindowsFileSystem_Tags
=== PAUSE TestAccAWSFsxWindowsFileSystem_Tags
=== RUN   TestAccAWSFsxWindowsFileSystem_ThroughputCapacity
=== PAUSE TestAccAWSFsxWindowsFileSystem_ThroughputCapacity
=== RUN   TestAccAWSFsxWindowsFileSystem_WeeklyMaintenanceStartTime
=== PAUSE TestAccAWSFsxWindowsFileSystem_WeeklyMaintenanceStartTime
=== RUN   TestAccAWSFsxWindowsFileSystem_auditConfig
=== PAUSE TestAccAWSFsxWindowsFileSystem_auditConfig
=== CONT  TestAccAWSFsxWindowsFileSystem_basic
=== CONT  TestAccAWSFsxWindowsFileSystem_SecurityGroupIds
=== CONT  TestAccAWSFsxWindowsFileSystem_ThroughputCapacity
=== CONT  TestAccAWSFsxWindowsFileSystem_Tags
=== CONT  TestAccAWSFsxWindowsFileSystem_auditConfig
=== CONT  TestAccAWSFsxWindowsFileSystem_disappears
=== CONT  TestAccAWSFsxWindowsFileSystem_storageTypeHdd
=== CONT  TestAccAWSFsxWindowsFileSystem_StorageCapacity
=== CONT  TestAccAWSFsxWindowsFileSystem_singleAz2
=== CONT  TestAccAWSFsxWindowsFileSystem_SelfManagedActiveDirectory_Username
=== CONT  TestAccAWSFsxWindowsFileSystem_SelfManagedActiveDirectory
=== CONT  TestAccAWSFsxWindowsFileSystem_multiAz
=== CONT  TestAccAWSFsxWindowsFileSystem_WeeklyMaintenanceStartTime
=== CONT  TestAccAWSFsxWindowsFileSystem_AutomaticBackupRetentionDays
=== CONT  TestAccAWSFsxWindowsFileSystem_KmsKeyId
=== CONT  TestAccAWSFsxWindowsFileSystem_DailyAutomaticBackupStartTime
=== CONT  TestAccAWSFsxWindowsFileSystem_CopyTagsToBackups
--- PASS: TestAccAWSFsxWindowsFileSystem_SelfManagedActiveDirectory (3264.42s)
--- PASS: TestAccAWSFsxWindowsFileSystem_disappears (3446.19s)
--- PASS: TestAccAWSFsxWindowsFileSystem_WeeklyMaintenanceStartTime (3550.71s)
--- PASS: TestAccAWSFsxWindowsFileSystem_basic (3597.00s)
--- PASS: TestAccAWSFsxWindowsFileSystem_DailyAutomaticBackupStartTime (3676.88s)
--- PASS: TestAccAWSFsxWindowsFileSystem_auditConfig (3689.29s)
--- PASS: TestAccAWSFsxWindowsFileSystem_AutomaticBackupRetentionDays (3690.70s)
--- PASS: TestAccAWSFsxWindowsFileSystem_StorageCapacity (3925.09s)
--- PASS: TestAccAWSFsxWindowsFileSystem_ThroughputCapacity (3949.41s)
--- PASS: TestAccAWSFsxWindowsFileSystem_Tags (3979.15s)
--- PASS: TestAccAWSFsxWindowsFileSystem_singleAz2 (4319.90s)
--- PASS: TestAccAWSFsxWindowsFileSystem_storageTypeHdd (4522.83s)
--- PASS: TestAccAWSFsxWindowsFileSystem_SelfManagedActiveDirectory_Username (4575.30s)
--- PASS: TestAccAWSFsxWindowsFileSystem_SecurityGroupIds (4884.61s)
--- PASS: TestAccAWSFsxWindowsFileSystem_multiAz (4937.93s)
--- PASS: TestAccAWSFsxWindowsFileSystem_KmsKeyId (5199.66s)
--- PASS: TestAccAWSFsxWindowsFileSystem_CopyTagsToBackups (5261.80s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	5265.611s
```

**Pull request merged**